### ISSUE TITLE
DALI Equivalent to PyTorch's transforms.RandomAffine (Custom labels, Unbalanced datasets?)

### ISSUE THREAD

**Issue by Wilann** [NVIDIA/DALI#2736]
I can't seem to find a DALI version of Torchvision's RandomAffine. Is there one, and if so, what is it called? 
If not, is there a combination of functions I can use to get the "random scale" functionality of RandomAffine?

**Comment by JanuszL** [comment]
Hi @Wilann,

DALI doesn't have explicit RandomAffine, but it has [`WarpAffine`](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.ops.WarpAffine) which can be feed by any transformation matrix. To create one you can combine multiple operations from the [`transforms`](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#module-nvidia.dali.ops.transforms) namespace. You can read more in [this tutorial](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/math/geometric_transforms.html), especially [this section](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/math/geometric_transforms.html#Transform-Gallery).

**Comment by Wilann** [comment]
Hey @JanuszL, 

Thanks so much for the links. A bit of a side question - I'm reading the **Combine Transforms** section where the example code is: 
```
pipe = dali.pipeline.Pipeline(6, 3, 0, seed = 1234)
with pipe:
    jpegs, _ = fn.file_reader(file_root=root_dir, files=image_files)
    images = fn.image_decoder(jpegs, device="mixed")

    size = encoded_images_sizes(jpegs)
    center = size / 2

    keypoints = fn.numpy_reader(file_root=root_dir, files=keypoint_files)
    mt = fn.transforms.rotation(angle = fn.random.uniform(range=(-45, 45)), center = center)
    mt = fn.transforms.translation(mt, offset = (400,0))
    images = fn.warp_affine(images, matrix = mt, fill_value=0, inverse_map=False)
    keypoints = fn.coord_transform(keypoints, MT = mt)
    pipe.set_outputs(images, keypoints)

pipe.build()
images, keypoints = pipe.run()
```

Is there any difference between putting the transformations in the `define_graph()` function (which is what I though what should be done according to the **Getting Started** section) or outside it, like they do here?

**Comment by JanuszL** [comment]
> Is there any difference between putting the transformations in the define_graph() function (which is what I though what should be done according to the Getting Started section) or outside it, like they do here?

This example is using new, [functional style API](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/functional_api.html). Each function you use has a snake case functional counterpart. We plan to make move to functional API as a primary one starting from 1.0, documentation rework is in progress (including the getting started section). The most recent, development in progress documentation can be [found]https://docs.nvidia.com/deeplearning/dali/master-user-guide/docs/index.html().

**Comment by Wilann** [comment]
In the **Geometric Transforms** section, there seems to be no custom pipeline. So basically, functional style API will use the default `Pipeline`, then all the transformations will go into the with `pipe: section`?

**Comment by Wilann** [comment]
Also, I have a use-case where I need custom labels, have unbalanced datasets, and need to sample from them individually. 
I also need to perform different transformations on different classes. Is this possible with DALI?

My current PyTorch code looks like this: 
Note: 
- `AnglesDataset` is the custom dataset where if a folder is named, let's say, "1. class_1_name", the label will be 1 
- `model_utils.BalancedBatchSchedulerSampler` samples the same number of images from each class
```
train_main_data = model_utils.AnglesDataset(train_main_dir, transform=train_main_transforms)
train_sub_data = model_utils.AnglesDataset(train_sub_dir, transform=train_sub_transforms)
train_dataset = ConcatDataset([train_main_data, train_sub_data])

train_loader = DataLoader(dataset=train_dataset,
                          sampler=model_utils.BalancedBatchSchedulerSampler(dataset=train_dataset,
                                                                            batch_size=train_batch_size),
                          batch_size=train_batch_size,
                          shuffle=False,
                          pin_memory=True,
                          num_workers=num_workers)
```

**Comment by JanuszL** [comment]
> In the Geometric Transforms section, there seems to be no custom pipeline. So basically, functional style API will use the default Pipeline, then all the transformations will go into the with pipe: section?

There is no need to have a custom pipeline in the functional API. All you need is:
- create pipeline
- define it:
```
with pipe:
```
- build and use

The main point is that you no longer need to create a boilerplate code with a custom pipeline class, as you just want to define your pipeline easily.

> Also, I have a use-case where I need custom labels, have unbalanced datasets, and need to sample from them individually.
I also need to perform different transformations on different classes. Is this possible with DALI?

You can try using [`FileReader`](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.ops.FileReader) and `files` argument. You can create an order of files for the whole training (do any bucketing and/or biasing) you want.
Regarding conditional transforms - you may try [Conditional-Like Execution and Masking](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/expressions/expr_conditional_and_masking.html).
